hibernate-core/src/main/java/org/hibernate/query/hql/internal/SemanticQueryBuilder.java

```diff
@@ -1498,12 +1500,24 @@ public SqmDynamicInstantiation<?> visitInstantiation(HqlParser.InstantiationCont
 		}
 
 		if ( !dynamicInstantiation.checkInstantiation( creationContext.getTypeConfiguration() ) ) {
-			final String typeName = dynamicInstantiation.getJavaType().getSimpleName();
+			final String qualifiedName = dynamicInstantiation.getJavaType().getName();
 			if ( dynamicInstantiation.isFullyAliased() ) {
-				throw new SemanticException( "Missing constructor or attributes for injection into type '" + typeName + "'", query );
+				throw new SemanticException("Missing constructor or attributes for injection into type '" + qualifiedName + "'" +
+						"Expected arguments are: " + dynamicInstantiation.argumentTypes(), query);
 			}
 			else {
-				throw new SemanticException( "Missing constructor for type '" + typeName  + "'", query );
+				List<InstantiationHelper.MismatchConstructorFieldPositionInfo> list = dynamicInstantiation.findMismatchConstructorFieldPositions(creationContext.getTypeConfiguration());
+
+				String conjecture = "";
+				if (list != null && list.isEmpty() == false) {
+					conjecture = " The presumed mismatch field position is: " +
+							list.stream()
+									.map(InstantiationHelper.MismatchConstructorFieldPositionInfo::toString)
+									.collect(Collectors.joining(","));
```

Comment:
You can use `StringHelper.join()` for this.

Reply:
I have finished the modification